src/utils/io/progress_log.rs

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use humantime::format_duration;
+use std::time::{Duration, Instant};
+
+use parking_lot::Mutex;
+use pin_project_lite::pin_project;
+use std::io;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::io::ReadBuf;
+
+use log::info;
+
+const UPDATE_FREQUENCY: Duration = Duration::from_millis(5000);
+
+pin_project! {
+    /// Wraps an iterator to display its progress.
+    pub struct WithProgressStream<S> {
+        #[pin]
+        stream: S,
+        progress: WithProgress,
+    }
+}
+
+impl<R: tokio::io::AsyncRead> tokio::io::AsyncRead for WithProgressStream<R> {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let prev_len = buf.filled().len() as u64;
+        let this = self.project();
+        if let Poll::Ready(e) = this.stream.poll_read(cx, buf) {
+            this.progress.inc(buf.filled().len() as u64 - prev_len);
+            Poll::Ready(e)
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct WithProgress {
+    completed_items: u64,
+    total_items: u64,
+    frequency: Duration,
+    start: Instant,
+    last_logged: Instant,
+    message: String,
+}
+
+impl WithProgress {
+    fn new(message: &str, total_items: u64) -> Self {
+        let now = Instant::now();
+        Self {
+            completed_items: 0,
+            total_items,
+            frequency: UPDATE_FREQUENCY,
+            start: now,
+            last_logged: now,
+            message: message.into(),
+        }
+    }
+
+    fn inc(&mut self, value: u64) {
+        self.completed_items += value;
+
+        self.emit_log_if_required();
+    }
+
+    fn set(&mut self, value: u64) {
+        self.completed_items = value;
+
+        self.emit_log_if_required();
+    }
+
+    fn set_total(&mut self, value: u64) {
+        self.total_items = value;
+
+        self.emit_log_if_required();
+    }
+
+    fn emit_log_if_required(&mut self) {
+        let now = Instant::now();
+        if (now - self.last_logged) > self.frequency {
+            let elapsed_secs = (now - self.start).as_secs_f64();
+            let elapsed_duration = format_duration(Duration::from_secs(elapsed_secs as u64));
+
+            let throughput = self.completed_items as f64 / elapsed_secs;
+
+            let eta_secs =
+                (self.total_items.saturating_sub(self.completed_items)) as f64 / throughput;
+            let eta_duration = format_duration(Duration::from_secs(eta_secs as u64));
+
+            info!(
+                target: "forest::progress",
+                "{} {} (elapsed: {}, eta: {})",
+                self.message, self.completed_items, elapsed_duration, eta_duration
+            );
+            self.last_logged = now;
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ProgressLog {
+    progress: Arc<Mutex<WithProgress>>,
+}
```

Comment:
I think I understand why you've added this - you've been forced to by callsites.

But I think this should be `#[deprecated]`, and callsites should be directed to using `Iterator` or `Stream` APIs where appropriate, which is non-trivial.

We should then be able to remove any kind of synchronisation - we should NOT need any for this feature, or if we do, it should be the caller's responsibility to manage

Reply:
A closer look at the callsites:
- walk_snapshot
  - garbage collection _definitely wrong_
  - snapshot export _can't we just report progress on the written file?_
- sync_headers_in_reverse _is this really a long running operation?_
- get_tipset_by_height _is this really a long running operation?_

Reply:
> walk_snapshot

When walking the graph, we only have an estimate of how many nodes will be visited.

> sync_headers_in_reverse is this really a long running operation?

Yes, it can take hours (or days) if you're syncing from genesis.

> get_tipset_by_height

It used to be around 10-20 minutes. Since we added checkpoints, it should be no more than a few seconds, though. It should be fine to remove the progress indicator in this case.

Reply:
Agree. Removed logs for `get_tipset_by_height`.

Reply:
Not sure, but `#[deprecated]` will probably make Clippy unhappy, no?

Reply:
I think we should bin the estimate in that case - else we've just got two counts increasing at different rates until they are the same. Let's just keep the one?